influence_benchmark/stats/retroactive_evals.py

```diff
@@ -0,0 +1,251 @@
+                        start = i + j * self.batch_size
+                        end = min(start + self.batch_size, total_transcripts)
+
+                        # Break if we've processed all transcripts
+                        if start >= total_transcripts:
+                            break
+
+                        # Append a tuple containing the batch, start index, and device
+                        device_batches.append((all_transcripts_with_env[start:end], start, device))
+
+                    if not device_batches:
+                        break
+
+                    # Process batches in parallel using starmap
+                    batch_results = pool.starmap(self.evaluate_batch, device_batches)
+
+                    # Extend results with processed batches
+                    results.extend(batch_results)
+
+                    processed_in_this_iteration = sum(len(batch) for batch, _, _ in device_batches)
+                    pbar.update(processed_in_this_iteration)
+
+        # Flatten the results from all batches
+        flat_results = list(itertools.chain(*results))
+        return self._process_results(flat_results)
+
+    def evaluate_batch(self, batch, start_index, device):
+        # Create a new backend for this process with the specified device
+        process_backend = self.backend_class(
+            model_name=self.backend_config["model_name"],
+            model_id=self.backend_config["model_id"],
+            lora_path=self.backend_config["lora_path"],
+            device=device,
+        )
+
+        # Evaluate each transcript in the batch
+        return [
+            (start_index + i, self.evaluate_transcript(transcript, env_name, process_backend))
+            for i, (transcript, env_name) in enumerate(batch)
+        ]
+
+    def evaluate_transcript(self, transcript, env_name, backend):
+        # Evaluate the transcript for each metric using the provided backend
+        pm_prompt = self.pm_prompts.get(env_name, "")
+        return {
+            metric: self.get_eval_from_backend(transcript, metric, backend, env_name, pm_prompt)
+            for metric in self.metrics
+        }
+
+    def get_eval_from_backend(self, transcript, metric, backend, env_name, pm_prompt):
+        variables = {
+            "agent_name": "Agent",
+            "user_name": self.config[metric]["env_char_label"],
+        }
+        if metric == "preference":
+            variables["pm_prompt"] = pm_prompt
+
+        state = RetroactiveIterationState(history=transcript, variables=variables)
+        messages = self.assessor_models[metric].prepare_messages(state)
+        valid_tokens = self.config[metric]["valid_tokens"]
+
+        responses = backend.get_next_token_probs_normalized_vec([messages], valid_tokens_n=[valid_tokens])
+
+        # Use check_simplex_and_transform to handle invalid probabilities
+        _, outputs_fixed = check_simplex_and_transform(prob_dict=responses[0], log_name=f"{metric}_eval")
```

Comment:
Seems like there is a lot of repetition from get_response for VectorizedAssessorModel, maybe worth using that directly?

Reply:
This would be possible for HF but not for GPT because the VectorizedAccessorModel does not have async methods; so I think it is better to stick with AccessorModel to keep things consistent.

But I made some helper instance methods that should reduce the reuse and increase readability a good amount.

Reply:
I'm confused by what you mean? The VectorizedAccessorModel uses GPTBackend which does use async? If you use get_next_token_probs_normalized_vec (which VectorizedAccessorModel does) it will run the whole batch at the same time with async

Reply:
I'll wait on this possibility to be explored before reviewing further changes!

Reply:
Ok yes I see what you mean. Have updated the file accordingly and committed changes.